src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java

```diff
@@ -96,8 +97,18 @@ public static SearchRequest parseSearchRequest(RestRequest request) {
             }
         }
 
+        // do not allow 'query_and_fetch' or 'dfs_query_and_fetch' search types
+        // from the REST layer. these modes are an internal optimization and should
+        // not be specified explicitly by the user.
+        String searchType = request.param("search_type");
+        if (SearchType.fromString(searchType).equals(SearchType.QUERY_AND_FETCH) ||
+                SearchType.fromString(searchType).equals(SearchType.DFS_QUERY_AND_FETCH)) {
+            throw new ElasticsearchIllegalArgumentException("Unsupported search type [" + searchType + "]");
```

Comment:
By the way I was just looking at call sites of `SearchType.fromString` and it looks to me that removing `(dfs_)query_and_fetch` from there would have the effect that you expect without having to add this special case to RestSearchAction? This way it would also have the good side effect of applying the same restriction to the `more_like_this` action.